runtime/commonTest/src/kotlinx/serialization/SerialDescriptorSpecificationTest.kt

```diff
@@ -0,0 +1,188 @@
+        run {
+            assertFalse(d.isElementOptional(0))
+            assertTrue(d.isElementOptional(1))
+            assertFalse(d.isElementOptional(2))
+        }
+        // Children descriptors
+        run {
+            val aDescriptor = d.getElementDescriptor(0)
+            assertTrue(aDescriptor.isNullable)
+            assertTrue(aDescriptor.kind is PrimitiveKind.INT)
+
+            val bDescriptor = d.getElementDescriptor(1)
+            assertFalse(bDescriptor.isNullable)
+            assertTrue(bDescriptor.kind is PrimitiveKind.STRING)
+
+            val cDescriptor = d.getElementDescriptor(2)
+            assertFalse(cDescriptor.isNullable)
+            assertTrue(cDescriptor.kind is PrimitiveKind.LONG)
+        }
+        // Failure modes
+        // TODO sandwraith initialization order assertFailsWith<IndexOutOfBoundsException> { d.isElementOptional(3) }
+        assertFailsWith<IndexOutOfBoundsException> { d.isElementOptional(Int.MAX_VALUE) }
+        assertFailsWith<IndexOutOfBoundsException> { d.getElementAnnotations(3) }
+        assertFailsWith<IndexOutOfBoundsException> { d.getElementName(3) }
+        assertFailsWith<IndexOutOfBoundsException> { d.getElementAnnotations(3) }
+    }
+
+    @SerialName("Named")
+    @Serializable
+    @Suppress("UNUSED")
+    enum class NamedEnum {
+        @SerialId(42)
+        FIRST,
+        @SerialName("SECOND")
+        THIRD
+    }
+
+    @Test
+    fun testEnumDescriptor() {
+        val d = NamedEnum.serializer().descriptor
+        assertEquals(UnionKind.ENUM_KIND, d.kind)
+        assertEquals("Named", d.serialName)
+        assertEquals(2, d.elementsCount)
+        assertFalse(d.isNullable)
+        // Names
+        assertEquals(0, d.getElementIndex("FIRST"))
+        assertEquals(1, d.getElementIndex("SECOND"))
+        assertEquals(UNKNOWN_NAME, d.getElementIndex("THIRD"))
+        // Elements
+        assertEquals("FIRST", d.getElementName(0))
+        assertEquals("SECOND", d.getElementName(1))
+        assertFailsWith<IndexOutOfBoundsException> { d.getElementName(2) }
+        // Annotations
+        d.assertSingleAnnotation(0) { it is SerialId && it.id == 42 }
+        assertEquals(0, d.getElementAnnotations(1).size)
+        // Element descriptors
+        assertSame(d, d.getElementDescriptor(0))
+        assertSame(d, d.getElementDescriptor(1))
+        assertFailsWith<IndexOutOfBoundsException> { d.getElementDescriptor(2) }
+        // Optionality
+        assertFailsWith<IllegalStateException> { d.isElementOptional(0) }
+        assertFailsWith<IllegalStateException> { d.isElementOptional(1) }
+    }
+
+    @Test
+    fun testListDescriptor() {
+        val descriptor = ArrayListSerializer(IntSerializer).descriptor
+        assertEquals("kotlin.collections.ArrayList", descriptor.serialName)
+        assertFalse(descriptor.isNullable)
+        assertEquals(1, descriptor.elementsCount)
+        assertSame(IntSerializer.descriptor, descriptor.getElementDescriptor(0))
```

Comment:
add `assertFailsWith<IllegalStateException> { d.isElementOptional(0) }` ? (and to `testMap`, too)

Reply:
Done. Though I am still not sure this is the right contract for maps/lists